More than 6 menu rows

Brycey92 opened the pull request:
Also:

- Add a control variable to disable different X offsets
- Get more values programmatically instead of hardcoding them

This PR reveals any `Back` and `Exit` menu rows that were previously not displayed due to being the 7th menu row.

~~All `meleeMenu_t` structs allocated will now cost 250 additional bytes.~~

Brycey92 commented:
This is required for the upcoming NVS manager, #359, as currently planned to be implemented

AEFeinstein commented:
Before doing the code part of the review, I compiled and looked at the UI. It seems to work, but I find it jarring. There's something about having all the offsets shift at once that makes my brain think it's jumped to a second page of options, rather than scroll every single one. I see two options to make this better:
1. Change scrolling behavior so that it scrolls a whole page (i.e. five items) at a time. For example, "Settings" would be
    * Page 1
        * Music
        * SFX
        * Screensaver
        * TFT
        * LED
    * Page 2
        * Mic Gain
        * Back
1. Make scrolling smooth by moving the items 1px at a time from start to destination. The menus on the Chainsaw scrolled this way

I like the arrow indicator for "there's more" a lot. You made the menu loop around, but the arrow is ever either on top or bottom. To keep it honest, either put arrows in both directions (valid scrolls), or lock the scroll at the top and bottom so it doesn't loop around.

AEFeinstein commented on emu/src/emu_sensors.c:
```diff
@@ -123,11 +129,13 @@ bool checkButtonQueue(buttonEvt_t* evt)
  */
 void emuSensorHandleKey( int keycode, int bDown )
 {
+#ifdef QUIT_EMU_KEY_ENABLED
```
What was this change for?

Brycey92 replied:
It was so I can retain my sanity when using screenshot tools, screen rulers, and magnifiers. `Alt + F4` or `Fn + Alt + F4` work just as well to exit the emu, but if someone really wants `Esc` to exit, they can still define the variable to turn it on.

AEFeinstein replied:
@dylwhich do you care?

Brycey92 replied:
![image](https://user-images.githubusercontent.com/666399/205517996-9c609c42-571a-4beb-9c7d-ac7456fb658a.png)

![image](https://user-images.githubusercontent.com/666399/205518027-30fce29d-9b7c-4bac-ad00-e73a8f6c3fba.png)

He can still weigh in, but I disabled it after he said I could do it if I wanted to

AEFeinstein commented on main/meleeMenu.h:
```diff
@@ -12,7 +12,8 @@
 
-#define MAX_ROWS 6
+#define MAX_ROWS 256
+#define MAX_ROWS_ON_SCREEN 6
```
If you allow scrolling, this should be 5 like the OG menu.
![image](https://user-images.githubusercontent.com/231180/205504934-b9c2e8e5-7c26-4e6a-be03-52fa06e74810.png)

Brycey92 replied:
I didn't want to mess up menus that had planned for 6 being the default, and I also didn't want to re-plan the spacing, but I can adjust if you prefer 5

AEFeinstein replied:
If you look far back enough, it was written for 5, and got bumped to 6 when something got big. I don't love how the last item overlaps the colored border, but 6 was easier for my sanity than whatever the alternative was at the time.

Brycey92 replied:
Oh, I didn't even notice that it wasn't supposed to overlap the border; thought that was some sort of asymmetrical design like the different row offsets.

So that's a "yes, switch to 5 rows", right?

Brycey92 replied:
I assumed that was a yes, and implemented it

AEFeinstein commented:
> All `meleeMenu_t` structs allocated will now cost 250 additional bytes.

Wrong, they will have an extra 250 `const char*`s. Pointers on a 32 bit processor are 4 bytes each, so it's 750 more bytes. That's pretty gratuitous. Please pick a more realistic number or change it to use dynamic memory allocation.

AEFeinstein commented on main/meleeMenu.h:
```diff
@@ -27,8 +28,10 @@ typedef struct
+    uint8_t firstRowOnScreen;
     uint8_t selectedRow;
     uint8_t allowLEDControl;
+    uint8_t usePerRowXOffsets;
```
if you're gonna use this like a `bool`, declare it as a `bool` and set it to `true` or `false`

AEFeinstein replied:
Also it's always set to `1`, so it doesn't seem particularly useful.

Brycey92 replied:
`allowLEDControl` is used as a bool but defined as a uint8_t so I standardized on that.
I do actually set `usePerRowXOffsets` to 0 in my `nvs-manager` branch

Brycey92 replied:
Because `allowLEDControl` is defined that way. Should both be defined as `bool` to match their usage?

Brycey92 replied:
Assumed yes, implemented

AEFeinstein replied:
Yes, I was bad and no one reviews me :(

AEFeinstein commented on main/meleeMenu.c:
```diff
@@ -113,7 +124,7 @@ void deinitMeleeMenu(meleeMenu_t* menu)
 {
     // Make sure there's space for this row
-    if(menu->numRows < MAX_ROWS)
+    if(menu->numRows < MAX_ROWS - 1)
```
Why did this get a `- 1`?

Brycey92 replied:
compiler deemed the `if` statement to always be true otherwise due to data type limitations. I'm just now realizing `numRows` is not a 0-indexed number, so `MAX_ROWS` will need to be 255 instead.

Brycey92 commented:
> I like the arrow indicator for "there's more" a lot. You made the menu loop around, but the arrow is ever either on top or bottom. To keep it honest, either put arrows in both directions (valid scrolls), or lock the scroll at the top and bottom so it doesn't loop around.

I had considered this dilemma, but my thinking was that the arrows should indicate "new" menu items (i.e. if someone were scrolling through options for the first time, they would know when they reached the end.) That would argue for locking the scroll at the top/bottom, but being able to jump between top and bottom is very nice, especially if there's something you often use at the bottom of a menu, or if you don't want to scroll through up to a theoretical 255 rows to get where you're going.

I could perhaps implement a jump-to-top/bottom icon or text or something, to show that it's something different than scrolling through rows one-by-one.

AEFeinstein commented:
I see your line of thinking, but the arrow doesn't indicate "new" things, it indicates "you can go here." A different icon isn't worth it. Just make everything consistent!

Brycey92 commented:
Hmm, would that mean an up arrow should always be drawn, since you can always go "up" to wrap around?

Brycey92 commented:
Actually, you can do the same thing on the bottom, so should arrows ALWAYS be displayed?

AEFeinstein commented:
> Hmm, would that mean an up arrow should always be drawn, since you can always go "up" to wrap around?
> Actually, you can do the same thing on the bottom, so should arrows ALWAYS be displayed?

yes and yes!

Brycey92 commented:
I switched on always showing the arrows, but with either of the methods for displaying arrows at topmost/bottommost entries, it's really unclear to me where the menu bounds are. I'm going to test using a slight change to the topmost/bottommost arrows later today (jeez, did I pull another all-nighter? 😅) to see if that feels more natural.

dylwhich commented:
One thing I'm noticing is now that the arrows are always shown, the odd-even fill for the menu items is broken. Though I do like how it looks with the arrows always showing underneath other than that.

Brycey92 commented:
Ah, I think there was a mis-documentation when I added `ALWAYS_SHOW_ARROWS`, since a more apt name might be `SHOW_ARROWS_EVEN_WHEN_ALL_THE_MENU_ROWS_CAN_FIT_ON_ONE_SCREEN` lol. Updated the code so they're hidden during animations, but I gotta think of a better name for that #define

Brycey92 commented:
Also, I made the wraparound arrows have bars to indicate they're something different from regular arrows. I basically took the Home/End symbols from my keyboard and turned them. Added some #defines for tweaking the arrow bars too. Give em a look!

I think we just need to get the animation reverse-direction edge cases fully sorted and then this'll be good to merge

AEFeinstein commented:
I still, just don't like it. It feels odd that if you keep tapping in one direction, the menu shifts in both directions. There's a lack of consistency. I also don't think that the arrows should be any different. If the menu wraps around, don't think about it with a start or an end. It's just a loop!

Looking at what's been done so far, my preference would be to keep having things fly in/out in the direction that the menu is scrolling if you scroll at an edge. So on the main menu when scroll past Credits, first Games flies out, then Tools flies out, then Music, etc.

Also I think you still have to resolve menus taking up an extra 1kb too.

Brycey92 commented:
I already did resolve the 1kb issue: https://github.com/AEFeinstein/Super-2023-Swadge-FW/pull/365/commits/ae64ce13839dafeb88569bdd2a2e57a39d05a91d

Brycey92 commented:
Since we both have very different opinions on how scrolling through menus should feel, maybe we should solicit more user feedback than our current sample size of 2 😛

AEFeinstein commented:
I stand corrected on the 1kb issue. Nicely done!

I'm all for soliciting more feedback on this. This must be what it's like when I think something is perfect and Dac unloads a bunch of design changes on me. Karma or something?

Brycey92 commented:
Why does this feel like it's rapidly becoming the Blue Ball Machine? 😂 https://blueballfixed.ytmnd.com/

Brycey92 commented:
I just tested in the NVS manager, and the time it takes to scroll through the entire menu to get to the opposite end when I go off one edge is way too long :sob:

dylwhich commented:
> I just tested in the NVS manager, and the time it takes to scroll through the entire menu to get to the opposite end when I go off one edge is way too long sob

Gotta bump up that `ANIM_MAXSPEED`!

AEFeinstein commented:
@Brycey92 to be clear, the onus is on you to bring these changes to dac and the team to see how they feel

Brycey92 commented on main/meleeMenu.c:
```diff
@@ -15,35 +16,57 @@
     {.r = 0x20, .g = 0x10, .b = 0x10},
     {.r = 0x00, .g = 0x20, .b = 0x10},
-    {.r = 0x20, .g = 0x20, .b = 0x10},
+    {.r = 0x20, .g = 0x20, .b = 0x00},
```
Ah, I remember making that change to try to get that color to display better at low brightness, but I must have forgotten I had left it in when you told me not to scale or change them. Will revert.

Brycey92 commented:
Closed in favor of #384
